File: ohara-client/src/test/scala/com/island/ohara/client/kafka/TestTopicAdmin.scala
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.client.kafka

import com.island.ohara.client.kafka.TopicAdmin.TopicInfo
import com.island.ohara.common.data.Serializer
import com.island.ohara.common.setting.TopicKey
import com.island.ohara.common.util.{CommonUtils, Releasable}
import com.island.ohara.kafka.Producer
import com.island.ohara.testing.With3Brokers
import org.apache.kafka.common.config.TopicConfig
import org.apache.kafka.common.errors.{InvalidPartitionsException, UnknownTopicOrPartitionException}
import org.junit.{After, Test}
import org.scalatest.Matchers

import scala.concurrent.ExecutionContext.Implicits.global
class TestTopicAdmin extends With3Brokers with Matchers {

  private[this] val topicAdmin = TopicAdmin(testUtil().brokersConnProps())
  private[this] val GROUP = "topic_group"

  private[this] def waitAndGetTopicInfo(topicKey: TopicKey): TopicInfo = {
    // wait the topic to be available
    CommonUtils.await(
      () =>
        try result(topicAdmin.topics().map(_.exists(_.name == topicKey.topicNameOnKafka())))
        catch {
          // the partition is not ready
          case _: NoSuchElementException           => false
          case _: UnknownTopicOrPartitionException => false
      },
      java.time.Duration.ofSeconds(60)
    )
    result(topicAdmin.topics()).find(_.name == topicKey.topicNameOnKafka()).get
  }

  @Test
  def createTopic(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val numberOfPartitions: Int = 1
    val numberOfReplications: Short = 1
    result(
      topicAdmin.creator
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicKey(topicKey)
        .create())
    val topic = waitAndGetTopicInfo(topicKey)
    topic.name shouldBe topicKey.topicNameOnKafka()
    topic.numberOfPartitions shouldBe numberOfPartitions
    topic.numberOfReplications shouldBe numberOfReplications

    result(topicAdmin.topics()).find(_.name == topicKey.topicNameOnKafka()).get shouldBe topic

    result(topicAdmin.delete(topicKey)) shouldBe true

    result(topicAdmin.topics()).find(_.name == topicKey.topicNameOnKafka()) shouldBe None
  }

  @Test
  def addPartitions(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val numberOfPartitions: Int = 1
    val numberOfReplications: Short = 1
    result(
      topicAdmin.creator
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicKey(topicKey)
        .create()
    )
    val topic = waitAndGetTopicInfo(topicKey)
    result(topicAdmin.changePartitions(topicKey, numberOfPartitions + 1))
    val topic2 = waitAndGetTopicInfo(topicKey)
    topic2.numberOfPartitions shouldBe topic.numberOfPartitions + 1
  }

  @Test
  def reducePartitions(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val numberOfPartitions: Int = 2
    val numberOfReplications: Short = 1
    result(
      topicAdmin.creator
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicKey(topicKey)
        .create())
    waitAndGetTopicInfo(topicKey)
    an[InvalidPartitionsException] should be thrownBy result(
      topicAdmin.changePartitions(topicKey, numberOfPartitions - 1))
  }

  @Test
  def negativePartitions(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val numberOfPartitions: Int = 2
    val numberOfReplications: Short = 1
    result(
      topicAdmin.creator
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicKey(topicKey)
        .create())
    waitAndGetTopicInfo(topicKey)
    an[InvalidPartitionsException] should be thrownBy result(topicAdmin.changePartitions(topicKey, -10))
  }

  @Test
  def keepPartitions(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val numberOfPartitions: Int = 2
    val numberOfReplications: Short = 1
    result(
      topicAdmin.creator
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicKey(topicKey)
        .create())
    waitAndGetTopicInfo(topicKey)
    an[InvalidPartitionsException] should be thrownBy result(topicAdmin.changePartitions(topicKey, numberOfPartitions))
  }

  @Test
  def changePartitionsOfNonexistentTopic(): Unit =
    an[UnknownTopicOrPartitionException] should be thrownBy result(
      topicAdmin.changePartitions(TopicKey.of(GROUP, CommonUtils.randomString(10)), 10))

  @Test
  def deleteNonexistentTopic(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    result(topicAdmin.delete(TopicKey.of(GROUP, CommonUtils.randomString()))) shouldBe false
    result(topicAdmin.creator.topicKey(topicKey).create())
    waitAndGetTopicInfo(topicKey)
    result(topicAdmin.delete(topicKey)) shouldBe true
  }

  @Test
  def testCleanupPolicy(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    result(topicAdmin.creator.topicKey(topicKey).cleanupPolicy(CleanupPolicy.DELETE).create())
    val topic = waitAndGetTopicInfo(topicKey)
    topic.configs(TopicConfig.CLEANUP_POLICY_CONFIG) shouldBe CleanupPolicy.DELETE.name
  }

  @Test
  def testCustomConfigs(): Unit = {
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    val key = TopicConfig.SEGMENT_BYTES_CONFIG
    val value = 1024 * 1024
    result(topicAdmin.creator.topicKey(topicKey).config(key, value.toString).create())
    val topic = waitAndGetTopicInfo(topicKey)
    topic.configs(key) shouldBe value.toString
  }

  @Test
  def testExist(): Unit = {
    result(topicAdmin.exist(TopicKey.of(GROUP, CommonUtils.randomString()))) shouldBe false
    val name = CommonUtils.randomString(10)
    val topicKey = TopicKey.of(GROUP, name)
    result(topicAdmin.creator.topicKey(topicKey).create())
    waitAndGetTopicInfo(topicKey)
    result(topicAdmin.exist(topicKey)) shouldBe true
  }

  @Test
  def testBeginningOffsetAndEndOffset(): Unit = {
    result(topicAdmin.exist(TopicKey.of(GROUP, CommonUtils.randomString()))) shouldBe false
    val topicKey = TopicKey.of(GROUP, CommonUtils.randomString(10))
    result(topicAdmin.creator.topicKey(topicKey).create())
    waitAndGetTopicInfo(topicKey)
    val producer = Producer.builder
      .connectionProps(testUtil().brokersConnProps())
      .keySerializer(Serializer.STRING)
      .valueSerializer(Serializer.STRING)
      .build()
    val count = 10
    try {
      (0 until count).foreach(
        _ =>
          producer
            .sender()
            .topicName(topicKey.topicNameOnKafka())
            .key(CommonUtils.randomString())
            .value(CommonUtils.randomString())
            .send())
      producer.flush()
    } finally Releasable.close(producer)
    val topicInfo = result(topicAdmin.topics()).find(_.name == topicKey.topicNameOnKafka()).get
    topicInfo.partitionInfos.foreach { partition =>
      partition.beginningOffset shouldBe 0
      partition.endOffset shouldBe count
    }
  }

  @After
  def tearDown(): Unit = Releasable.close(topicAdmin)
}

File: ohara-kafka/src/test/java/com/island/ohara/kafka/TestTopicAdmin.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.kafka;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

import com.island.ohara.common.util.CommonUtils;
import com.island.ohara.common.util.Releasable;
import com.island.ohara.testing.With3Brokers;
import java.time.Duration;
import java.util.Collections;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import org.apache.kafka.common.config.TopicConfig;
import org.junit.After;
import org.junit.Test;

public class TestTopicAdmin extends With3Brokers {
  private final TopicAdmin client = TopicAdmin.of(testUtil().brokersConnProps());

  private void waitPartitions(String topicName, int numberOfPartitions) {
    CommonUtils.await(
        () -> {
          try {
            return client
                    .topicDescription(topicName)
                    .toCompletableFuture()
                    .get()
                    .numberOfPartitions()
                == numberOfPartitions;
          } catch (Exception e) {
            return false;
          }
        },
        Duration.ofSeconds(30));
  }

  @Test
  public void testAddPartitions() throws ExecutionException, InterruptedException {
    String topicName = CommonUtils.randomString(10);
    int numberOfPartitions = 1;
    short numberOfReplications = 1;
    client
        .topicCreator()
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicName(topicName)
        .create()
        .toCompletableFuture()
        .get();
    waitPartitions(topicName, numberOfPartitions);
    assertEquals(
        numberOfPartitions,
        client.topicDescription(topicName).toCompletableFuture().get().numberOfPartitions());

    numberOfPartitions = 2;
    client.createPartitions(topicName, numberOfPartitions).toCompletableFuture().get();
    waitPartitions(topicName, numberOfPartitions);
    assertEquals(
        numberOfPartitions,
        client.topicDescription(topicName).toCompletableFuture().get().numberOfPartitions());
    // decrease the number
    assertException(
        Exception.class, () -> client.createPartitions(topicName, 1).toCompletableFuture().get());
    // alter an nonexistent topic
    assertException(
        NoSuchElementException.class,
        () -> client.createPartitions("Xxx", 2).toCompletableFuture().get(),
        true);
  }

  @Test
  public void testCreate() throws ExecutionException, InterruptedException {
    String topicName = CommonUtils.randomString(10);
    int numberOfPartitions = 2;
    short numberOfReplications = (short) 2;
    client
        .topicCreator()
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .topicName(topicName)
        .create()
        .toCompletableFuture()
        .get();
    waitPartitions(topicName, numberOfPartitions);
    TopicDescription topicInfo = client.topicDescription(topicName).toCompletableFuture().get();

    assertEquals(topicName, topicInfo.name());

    assertEquals(numberOfPartitions, topicInfo.numberOfPartitions());
    assertEquals(numberOfReplications, topicInfo.numberOfReplications());

    assertEquals(
        topicInfo,
        client.topicDescriptions().toCompletableFuture().get().stream()
            .filter(t -> t.name().equals(topicName))
            .findFirst()
            .get());

    client.deleteTopic(topicName).toCompletableFuture().get();
    assertFalse(client.exist(topicName).toCompletableFuture().get());
  }

  @Test
  public void testTopicOptions() throws ExecutionException, InterruptedException {
    String topicName = CommonUtils.randomString(10);
    int numberOfPartitions = 2;
    short numberOfReplications = (short) 2;
    Map<String, String> options =
        Collections.singletonMap(
            TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_DELETE);
    client
        .topicCreator()
        .numberOfPartitions(numberOfPartitions)
        .numberOfReplications(numberOfReplications)
        .options(options)
        .topicName(topicName)
        .create()
        .toCompletableFuture()
        .get();
    waitPartitions(topicName, numberOfPartitions);

    TopicDescription topicInfo = client.topicDescription(topicName).toCompletableFuture().get();

    assertEquals(topicName, topicInfo.name());

    assertEquals(numberOfPartitions, topicInfo.numberOfPartitions());
    assertEquals(numberOfReplications, topicInfo.numberOfReplications());

    assertEquals(
        TopicConfig.CLEANUP_POLICY_DELETE,
        topicInfo.options().stream()
            .filter(x -> Objects.equals(x.key(), TopicConfig.CLEANUP_POLICY_CONFIG))
            .collect(Collectors.toList())
            .get(0)
            .value());
  }

  @After
  public void cleanup() throws ExecutionException, InterruptedException {
    client
        .topicDescriptions()
        .toCompletableFuture()
        .get()
        .forEach(
            t -> {
              try {
                client.deleteTopic(t.name()).toCompletableFuture().get();
              } catch (Exception e) {
                throw new RuntimeException(e);
              }
            });
    Releasable.close(client);
  }
}
